File: backend/internal/model/db/projection/audit.go
package projection

import (
	"akatengu/internal/enums"
	"akatengu/internal/enums/event_types"
	"encoding/json"
)

// All audit types use manual mapping in the repo layer (no dbmap annotation).
// ExchangeRate is already defined in investment.go — reused here.

type AggregateVersionAudit struct {
	AggregateType  enums.AggregateType `json:"aggregate_type"`
	AggregateId    string              `json:"aggregate_id"`
	MerchantID     int64               `json:"merchant_id"`
	CurrentVersion int64               `json:"current_version"`
}

type EventStoreAudit struct {
	EventId          int64                 `json:"event_id"`
	EventUuid        string                `json:"event_uuid"`
	OccurredAt       string                `json:"occurred_at"`
	MerchantID       int64                 `json:"merchant_id"`
	AggregateType    enums.AggregateType   `json:"aggregate_type"`
	AggregateId      string                `json:"aggregate_id"`
	AggregateVersion int64                 `json:"aggregate_version"`
	EventType        event_types.EventType `json:"event_type"`
	Payload          json.RawMessage       `json:"payload"`
	Metadata         *json.RawMessage      `json:"metadata"`
	UpdatedBy        *string               `json:"updated_by"`
}

type CheckpointAudit struct {
	ProjectionName string  `json:"projection_name"`
	MerchantID     int64   `json:"merchant_id"`
	LastEventID    int64   `json:"last_event_id"`
	UpdatedAt      *string `json:"updated_at"`
}

type SnapshotAudit struct {
	SnapshotId    int64               `json:"snapshot_id"`
	MerchantID    int64               `json:"merchant_id"`
	AggregateType enums.AggregateType `json:"aggregate_type"`
	AggregateId   string              `json:"aggregate_id"`
	AtVersion     int64               `json:"at_version"`
	State         json.RawMessage     `json:"state"`
	CreatedAt     string              `json:"created_at"`
}

File: frontside/src/types/audit.ts
export interface AggregateVersionAudit {
  aggregate_type:  string;
  aggregate_id:    string;
  merchant_id:     number;
  current_version: number;
}

export interface EventStoreAudit {
  event_id:          number;
  event_uuid:        string;
  occurred_at:       string;
  merchant_id:       number;
  aggregate_type:    string;
  aggregate_id:      string;
  aggregate_version: number;
  event_type:        string;
  payload:           unknown;
  metadata:          unknown;
  updated_by:        string | null;
}

export interface CheckpointAudit {
  projection_name: string;
  merchant_id:     number;
  last_event_id:   number;
  updated_at:      string | null;
}

export interface SnapshotAudit {
  snapshot_id:    number;
  merchant_id:    number;
  aggregate_type: string;
  aggregate_id:   string;
  at_version:     number;
  state:          unknown;
  created_at:     string;
}

export interface ExchangeRate {
  rate_id:   number;
  currency:  string;
  rate_date: string;
  rate_twd:  string;
  source:    string;
}
